lib/components/session/login.vue

```diff
@@ -0,0 +1,79 @@
+<!--
+Copyright 2017 Super Adventure Developers
+See the NOTICE file at the top-level directory of this distribution and at
+https://github.com/nafundi/super-adventure/blob/master/NOTICE.
+
+This file is part of Super Adventure. It is subject to the license terms in
+the LICENSE file found in the top-level directory of this distribution and at
+https://www.apache.org/licenses/LICENSE-2.0. No part of Super Adventure,
+including this file, may be copied, modified, propagated, or distributed
+except according to the terms contained in the LICENSE file.
+-->
+<template>
+  <div>
+    <heading title="Log in"/>
+    <alert type="danger" :message="error"/>
+    <form @submit.prevent="logIn">
+      <div class="form-group">
+        <label for="email">Email address</label>
+        <input type="email" v-model="email" ref="email" id="email"
+          class="form-control" placeholder="Email" required>
+      </div>
+      <div class="form-group">
+        <label for="password">Password</label>
+        <input type="password" v-model="password" id="password"
+          class="form-control" placeholder="Password" required>
+      </div>
+      <button type="submit" class="btn btn-primary" :disabled="disabled">
+        Log in
+      </button>
+    </form>
+  </div>
+</template>
+
+<script>
+import axios from 'axios';
+
+import { safeBack } from '../../router';
+
+export default {
+  data() {
+    return {
+      email: '',
+      password: '',
+      error: null,
+      disabled: false
+    };
+  },
+  mounted() {
+    this.$refs.email.focus();
+  },
+  methods: {
+    logIn() {
+      this.disabled = true;
```

Comment:
similarly, i worry about this flag getting set/unset each time a form exists.

at the least, is there a `.complete` or something in axios so that you don't have to apply it separately to the success and fail branches?

but ideally, is there some way we can have the behaviour automatically apply?

if the answers are not obvious then we'll punt until they are.

Reply:
> at the least, is there a `.complete` or something in axios so that you don't have to apply it separately to the success and fail branches?

axios just returns a `Promise`. I do wish that standard `Promise`s had a `finally()` method. It looks like there's a stage-3 [candidate for that method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally). Right now our Babel configuration is set to support stage-2 candidates, but not stage-3. However, it'd be easy to change that setting so that we have access to `finally()`.

> but ideally, is there some way we can have the behaviour automatically apply?
>
> if the answers are not obvious then we'll punt until they are.

Maybe we could write a fetch function that looks for a `disabled` property on the component. One option is to include the function as a method on the component using [Vue mixins](https://vuejs.org/v2/guide/mixins.html). But I'm not totally sure yet what the right answer is. I would opt for waiting until we have more forms, then seeing whether a refactor becomes clearer.

Reply:
sounds like we should punt this one then.